llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp

```diff
@@ -0,0 +1,46 @@
+//===- CleanupSYCLCompilerInternalMetadata.cpp - remove SYCL compiler MD --===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Cleanup SYCL compiler internal metadata (usually inserted by sycl-post-link)
+// as it will never be used in the compilation ever again
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h"
+
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace {
+
+void cleanupSYCLCompilerModuleMetadata(const Module &M, llvm::StringRef MD) {
+  NamedMDNode *Node = M.getNamedMetadata(MD);
+  if (!Node)
+    return;
+  Node->clearOperands();
+  Node->dropAllReferences();
+  Node->eraseFromParent();
+}
+
+} // anonymous namespace
+
+PreservedAnalyses
+CleanupSYCLCompilerInternalMetadataPass::run(Module &M,
+                                             ModuleAnalysisManager &MAM) {
+  // Remove SYCL module-level metadata that will never be used again to avoid
+  // its duplication of their operands during llvm-link hence preventing
+  // increase of the module size
+  llvm::SmallVector<llvm::StringRef, 2> ModuleMDToRemove = {
+      "sycl_aspects", "sycl_types_that_use_aspects"};
+  for (const auto &MD : ModuleMDToRemove)
+    cleanupSYCLCompilerModuleMetadata(M, MD);
+
+  return PreservedAnalyses::all();
```

Comment:
I'm not an expert here but usually I see we keep track of if we actually did anything, and if so, we return `PreservedAnalyses::none()`, and if we did nothing, we return `PreservedAnalyses::all()`. In this case we only change metadata so I'm not sure if that is guaranteed to not modify any analyses.

Reply:
Thanks, I'm neither an expert, let me double check

Reply:
> I'm not an expert here but usually I see we keep track of if we actually did anything, and if so, we return `PreservedAnalyses::none()`, and if we did nothing, we return `PreservedAnalyses::all()`.

Note: this is a very pessimistic (and probably lazy) approach, which should ideally be avoided. In this case I think it is safe to say that we have preserved all standard LLVM analyses, because we are only removing metadata which we have inserted ourselves, we don't even touch any instructions.

Reply:
Thanks, I always saw that but had no idea if it was actually the right thing to do, good to know it's usually not :)